Back-deploy @objc actor types.

DougGregor opened the pull request:
@objc actors implicitly inherit from the new, hidden
`SwiftNativeNSObject` class that inherits from `NSObject` yet provides
Swift-native reference counting, which is important for the actor
runtime's handling of zombies. However, `SwiftNativeNSObject` is only
available in the Swift runtime in newer OS versions (e.g., macOS
12.0/iOS 15.0), and is available in the back-deployed _Concurrency
library, but there is no stable place to link against for
back-deployed code. Tricky, tricky.

When back-deploying @objc actors, record `NSObject` as the superclass
in the metadata in the binary, because we cannot reference
`SwiftNativeNSObject`. Then, emit a static initializer to
dynamically look up `SwiftNativeNSObject` by name (which will find it
in either the back-deployment library, on older systems, or in the
runtime for newer systems), then swizzle that in as the superclass of
the @objc actor.

Fixes rdar://83919973.

DougGregor committed:
Back-deploy @objc actor types.

@objc actors implicitly inherit from the new, hidden
`SwiftNativeNSObject` class that inherits from `NSObject` yet provides
Swift-native reference counting, which is important for the actor
runtime's handling of zombies. However, `SwiftNativeNSObject` is only
available in the Swift runtime in newer OS versions (e.g., macOS
12.0/iOS 15.0), and is available in the back-deployed _Concurrency
library, but there is no stable place to link against for
back-deployed code. Tricky, tricky.

When back-deploying @objc actors, record `NSObject` as the superclass
in the metadata in the binary, because we cannot reference
`SwiftNativeNSObject`. Then, emit a static initializer to
dynamically look up `SwiftNativeNSObject` by name (which will find it
in either the back-deployment library, on older systems, or in the
runtime for newer systems), then swizzle that in as the superclass of
the @objc actor.

Fixes rdar://83919973.

DougGregor commented:
@swift-ci please test

slavapestov commented:
If the @objc actor can be declared as a class_stub on the Clang side, you should be able to get rid of the static initializer. The class stub realization function should be able to set the superclass.

DougGregor committed:
Simplify @objc actor back-deploy IR and switch to `objc_getRequiredClass`.

`objc_getRequiredClass` will produce a fatal error if the class isn't
found, which will prevent a malformed program using back-deployed @objc
actor from launching. Also eliminate the spurious `objc_opt_self`
call, which is unneeded given that we're realizing the metadata.

Thanks to Mike Ash for the review.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
> If the @objc actor can be declared as a class_stub on the Clang side, you should be able to get rid of the static initializer. The class stub realization function should be able to set the superclass.

The static-initializers solution is so simple, and given this is a back-deployment scenario, I'd prefer to keep it simple and isolated rather than tie into the stub machinery. If it becomes a problem, we do have the option to change the mechanism later.
